lib/rules/no-undef-directives.js

```diff
@@ -0,0 +1,217 @@
+  // allow case-insensitive only when the directive name itself contains capitalized letters
+  for (const name of definedNames) {
+    if (
+      name.toLowerCase() === camelName.toLowerCase() &&
+      name !== name.toLowerCase()
+    ) {
+      return true
+    }
+  }
+
+  return false
+}
+
+module.exports = {
+  meta: {
+    type: 'problem',
+    docs: {
+      description: 'disallow use of undefined custom directives',
+      categories: undefined,
+      url: 'https://eslint.vuejs.org/rules/no-undef-directives.html'
+    },
+    fixable: null,
+    schema: [
+      {
+        type: 'object',
+        properties: {
+          ignore: {
+            type: 'array',
+            items: { type: 'string' },
+            uniqueItems: true
+          }
+        },
+        additionalProperties: true
+      }
+    ],
+    messages: {
+      undef: "The 'v-{{name}}' directive has been used, but not defined."
+    }
+  },
+  /** @param {RuleContext} context */
+  create(context) {
+    const options = context.options[0] || {}
+    const { ignore = [] } = options
+    const isAnyIgnored = regexp.toRegExpGroupMatcher(ignore)
+
+    /**
+     * Check whether the given directive name is a verify target or not.
+     *
+     * @param {string} rawName The directive name.
+     * @returns {boolean}
+     */
+    function isVerifyTargetDirective(rawName) {
+      const kebabName = casing.kebabCase(rawName)
+      if (
+        utils.isBuiltInDirectiveName(rawName) ||
+        isAnyIgnored(rawName, kebabName)
+      ) {
+        return false
+      }
+      return true
+    }
+
+    /**
+     * @param {(rawName: string) => boolean} isDefined
+     * @returns {TemplateListener}
+     */
+    function createTemplateBodyVisitor(isDefined) {
+      return {
+        /** @param {VDirective} node */
+        'VAttribute[directive=true]'(node) {
+          const name = node.key.name.name
+          if (utils.isBuiltInDirectiveName(name)) {
+            return
+          }
+          const rawName = node.key.name.rawName || name
+          if (isVerifyTargetDirective(rawName) && !isDefined(rawName)) {
+            context.report({
+              node: node.key,
+              messageId: 'undef',
+              data: {
+                name: rawName
+              }
+            })
+          }
+        }
+      }
+    }
+
+    if (utils.isScriptSetup(context)) {
+      // For <script setup>
+      /** @type {Set<string>} */
+      const definedInSetupDirectives = new Set()
+      /** @type {Set<string>} */
+      const definedInOptionDirectives = new Set()
+
+      const globalScope = context.sourceCode.scopeManager.globalScope
+      if (globalScope) {
+        for (const variable of globalScope.variables) {
+          definedInSetupDirectives.add(variable.name)
+        }
+        const moduleScope = globalScope.childScopes.find(
+          (scope) => scope.type === 'module'
+        )
+        for (const variable of (moduleScope && moduleScope.variables) || []) {
```

Comment:
I think this is a little more concise:

```suggestion
        for (const variable of moduleScope?.variables ?? []) {
```